### PR TITLE
Allow TreeItem nodes to toggle visibility

### LINKED ISSUE #4309: Add functionality to toggle visibility of TreeItems within a Tree
### Describe the project you are working on

An audio editor/loader to load a common game audio pack format (wwise .bnk files). As part of this I am populating a tree with potentially 1000's of rows of data.

### Describe the problem or limitation you are having in your project

To improve usability of my application I'd like to be able to filter the TreeItem's within the Tree so that it's easy to find the particular audio file or event the user wants. I was unable to find any easy way of filtering the entire tree without removing then re-adding nodes which was a little slow, as it required re-populating the entire tree every time just to keep everything in order.
For large trees this becomes very cumbersome.
Further I have a second tree which has links to the first (ie. clicking a row in one tree will select a row in the other tree). This is essentially impossible with the current restrictions as the referenced TreeItem will no longer exist if I filter the first tree.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

I propose a property and associated getter and setters:

- bool **visible**

| *Setter* | set_visible(value) |
| -------- | ------------------- |
| *Getter* | is_visible() |

If `true` then the TreeItem will not be drawn in the tree.
The TreeItem will also not be returned by the `get_next_visible` method and similar.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

I have done some basic work on trying to implement this, and the initial work to simply add the above methods and have it be visible or not works well and quickly. I am able to create a filterable tree very easily and it feels responsive even with 1000's of elements.

The getters/setters.
```cpp
void TreeItem::set_visible(bool p_visible) {
	if (visible == p_visible || !tree) {
		return;
	}
	visible = p_visible;
	tree->update();

	_changed_notify();
}

bool TreeItem::is_visible() {
	return visible;
}
```

Using this is as simple as updating a line in the tree.cpp::Tree::draw_item method:

```cpp
bool skip = ((p_item == root && hide_root) || !p_item->is_visible());
```

A number of other functions will need to be updated (I am still working on this) such as:

```cpp
int Tree::compute_item_height(TreeItem *p_item) const {
	if ((p_item == root && hide_root) || !p_item->is_visible()) {
		return 0;
	}
...
```

### If this enhancement will not be used often, can it be worked around with a few lines of script?

As far as I could find, there isn't really a good work around for this (especially in godot 3).
In Godot 4 the new functionality to move a TreeItem from one Tree to another could be kind of made to work, but again, it is probably slower than this and also will make it hard to keep the tree in order when filtering it.
Certainly not "a few lines of script".

### Is there a reason why this should be core and not an add-on in the asset library?

This will require changes right in the TreeItem and Tree functions, so I don't think it's possible to implement this functionality with an add-on.

There are two other propsals which I would consider related to this one:
https://github.com/godotengine/godot-proposals/issues/2088
https://github.com/godotengine/godot-proposals/issues/3607
The first is probably closer but doesn't actually cover this specific functionality I would like here.

I also think this should be in the core as it will allow for the concept of "filterable trees" to be implemented quickly and easily (as part of my testing I have loaded in a subset of the English dictionary and have a simple search bar at the top to filter by words. This however could be changed to filter by metadata or anything else associated with the row.

### PR DESCRIPTION
**Pull request by monkeyman192**
closes https://github.com/godotengine/godot-proposals/issues/4309

I have attached an example godot project (it's pretty lazy and dirty without some optimisations but it's simply to show how the functionality works).

~[filtered_tree.zip](https://github.com/godotengine/godot/files/8456769/filtered_tree.zip)~
[filtered_tree2.zip](https://github.com/godotengine/godot/files/8602910/filtered_tree2.zip)

As far as I can see for testing, the functionality works really nicely. One thing that may be confusing for people using the property can be explained by an example:

Let's say I have a Tree with TreeItem's as follows (assuming a hidden root, and everything is current expanded):

Fruits
├─ apple
└─ pear
Vegetables
└─ cucumber

If I set `apple` to be invisible then it will not be drawn, and everything else will.
If I then set `Fruits` to be invisble then that branch will not be drawn.
If I then set `Fruits` to be visible I'd expect it to be drawn again and same with `pear`, but not apple.
This is how it will currently work, however, if I checked `pear.visible` when `Fruits.visible = false`, then it will be `true` since I am not propagating the value to children (so that if a parent TreeItem is made visible again, it doesn't cause unintentded changes of state of child TreeItems.

This could be slightly confusing to users, however it could be potentially clarified by documentation, or potentially have the `is_visible()` getter return a computed value based on the visibility of the parent(s).
Very open to discussion and suggestions regarding these points 😄

### CODE REVIEW

**Comment by Calinou** [comment]
The editor's node tree filtering could be modified to make use of this new Tree visibility. It will likely speed up filtering large scene trees.

**Comment by KoBeWi** [comment]
I get crash with this code:
```GDScript
func _ready() -> void:
	var tree = $Tree as Tree
	
	var root = tree.create_item()
	root.set_text(0, "Test0")
	
	for i in 8:
		var item = tree.create_item(root)
		item.set_text(0, "Test1")
	
	for i in 4:
		var item = root.get_first_child().create_child()
		item.set_text(0, "Test2")
	
	root.get_first_child().visible = false
```
```
[0] TreeItem::get_parent (C:\godot_source\scene\gui\tree.cpp:656)
[1] Tree::get_column_minimum_width (C:\godot_source\scene\gui\tree.cpp:4207)
[2] Tree::get_internal_min_size (C:\godot_source\scene\gui\tree.cpp:3647)
[3] Tree::_do_incr_search (C:\godot_source\scene\gui\tree.cpp:3673)
[4] Tree::_notification (C:\godot_source\scene\gui\tree.cpp:3818)
[5] Tree::_notificationv (C:\godot_source\scene\gui\tree.h:365)
[6] Object::notification (C:\godot_source\core\object\object.cpp:850)
[7] CanvasItem::_update_callback (C:\godot_source\scene\main\canvas_item.cpp:143)
[8] call_with_variant_args_helper<CanvasItem> (C:\godot_source\core\variant\binder_common.h:241)
[9] call_with_variant_args_dv<CanvasItem> (C:\godot_source\core\variant\binder_common.h:384)
[10] MethodBindT<CanvasItem>::call (C:\godot_source\core\object\method_bind.h:336)
[11] Object::callp (C:\godot_source\core\object\object.cpp:839)
[12] Callable::call (C:\godot_source\core\variant\callable.cpp:62)
[13] MessageQueue::_call_function (C:\godot_source\core\object\message_queue.cpp:230)
[14] MessageQueue::flush (C:\godot_source\core\object\message_queue.cpp:277)
[15] SceneTree::physics_process (C:\godot_source\scene\main\scene_tree.cpp:421)
[16] Main::iteration (C:\godot_source\main\main.cpp:2692)
[17] OS_Windows::run (C:\godot_source\platform\windows\os_windows.cpp:677)
[18] widechar_main (C:\godot_source\platform\windows\godot_windows.cpp:175)
[19] _main (C:\godot_source\platform\windows\godot_windows.cpp:197)
[20] main (C:\godot_source\platform\windows\godot_windows.cpp:211)
[21] WinMain (C:\godot_source\platform\windows\godot_windows.cpp:225)
[22] __scrt_common_main_seh (D:\a01\_work\26\s\src\vctools\crt\vcstartup\src\startup\exe_common.inl:288)
[23] BaseThreadInitThunk
-- END OF BACKTRACE --
```
Caused by the last line.

File: scene/gui/tree.cpp [hunk 1]
```diff
@@ -543,6 +543,20 @@ bool TreeItem::is_collapsed() {
 	return collapsed;
 }
 
+void TreeItem::set_visible(bool p_visible) {
+	if (visible == p_visible || !tree) {
```
Comment (KoBeWi) [inline scene/gui/tree.cpp:547]: `!tree` shouldn't be here, because it makes it impossible to set the property when not inside tree.

Better put it after `visible =`.

**Comment by monkeyman192** [comment]
> I get crash with this code:
> 
> ```gdscript
> func _ready() -> void:
> 	var tree = $Tree as Tree
> 	
> 	var root = tree.create_item()
> 	root.set_text(0, "Test0")
> 	
> 	for i in 8:
> 		var item = tree.create_item(root)
> 		item.set_text(0, "Test1")
> 	
> 	for i in 4:
> 		var item = root.get_first_child().create_child()
> 		item.set_text(0, "Test2")
> 	
> 	root.get_first_child().visible = false
> ```

Thanks for finding this! I have fixed this issue but it made me realise another issue which I'll need to fix. When the parent `TreeItem` is set as `visible = false`, if it isn't collapsed then the child `TreeItem`'s are still visible (and actually get rendered in the same "level" as the parent that is no longer visible. Will work on fixing this.

**Comment by monkeyman192** [comment]
Ok. I have made a fix for the above issues mentioned.
As part of this I have also updated the example project (initial comment has been updated to include the new version)

**Comment by KoBeWi** [comment]
Folding is still active when all children all invisible:
godot windows tools 64_G3riW0cCQG
Is this intended?

File: scene/gui/tree.cpp [hunk 2]
```diff
@@ -544,6 +544,28 @@ bool TreeItem::is_collapsed() {
...
+bool TreeItem::is_visible() {
+	// Visibility is dependent on the visibility of parents.
+	bool is_visible = visible;
+	TreeItem *parent_ = parent;
+	while (parent_ && is_visible) {
+		is_visible &= parent_->visible;
+		parent_ = parent_->parent;
+	}
+	return is_visible;
+}
```
Comment (KoBeWi) [inline scene/gui/tree.cpp:567]: This is a setter for a variable, so it should only return that variable.
If you want function like this one here, it should be called `is_visible_in_tree()` or similar (like in CanvasItem).

File: scene/gui/tree.cpp [hunk 3]
```diff
@@ -682,7 +704,21 @@ TreeItem *TreeItem::get_prev_visible(bool p_wrap) {
...
+	while (prev && !prev->is_visible()) {
+		prev = prev->_get_prev_visible(p_wrap);
+		if (prev == loop) {
+			// Check that we haven't looped all the way around to the start.
+			prev = nullptr;
+			break;
+		}
+	}
+	return prev;
+}
```
Comment (KoBeWi) [inline scene/gui/tree.cpp:719]: I think it's better to integrate it into the original method

**Comment by monkeyman192** [comment]
> Folding is still active when all children all invisible:
> godot windows tools 64_G3riW0cCQG
> Is this intended?

Ah, good catch. I think it should not be.

> This is a setter for a variable, so it should only return that variable. If you want function like this one here, it should be called `is_visible_in_tree()` or similar (like in CanvasItem).
> 
> EDIT: Ah, but also you are traversing the whole tree upwards. This basically kills any performance optimizations you tried to achieve.

I assume you mean `getter`? I thought that in general there was nothing against having a getter return a modified version of the variable? Although it would lead to cases where you can do `ti.visible = true; print(ti.visible) # false` if the parent is not visible.
I think `is_visible_in_tree()` is a better idea though.
The main issue is that the `visible` property doesn't always actually reflect whether it will be visible as it depends on the parent (as I recently added).
The upward traversal does at least stop when it is found to not be visible. I guess the performance hit would be more when the tree is deeply nested. I'll see if there is a better away around this.

**Comment by KoBeWi** [comment]
> I assume you mean getter?

Ah, yes.

> The main issue is that the visible property doesn't always actually reflect whether it will be visible as it depends on the parent (as I recently added).

The same behavior is in e.g. Node2D. If it's `visible`, it doesn't mean you can see it, as the parent might be invisible. This is perfectly fine, because the node can exist outside of tree and the property needs to be independent.

> I guess the performance hit would be more when the tree is deeply nested.

Yes. But even with shallow tree, if you have e.g. 1000 items on 2 levels, each of the items would check parent and grandparent visibility needlessly every time you update the tree. `is_visible()` will be called very often, so it needs to be performant.

**Comment by monkeyman192** [comment]
> The same behaviour is in e.g. Node2D. If it's `visible`, it doesn't mean you can see it, as the parent might be invisible. This is perfectly fine, because the node can exist outside of tree and the property needs to be independent.

Ok cool. If there is precedent for this behaviour and it's understood, then I'm happy with that. I was just concerned because I felt it seemed unintuitive at first glance.

> Yes. But even with shallow tree, if you have e.g. 1000 items on 2 levels, each of the items would check parent and grandparent visibility needlessly every time you update the tree. `is_visible_in_tree()` will be called very often, so it needs to be performant.

Yep, very true. I'll work on this more tonight and see what I can do.

File: scene/gui/tree.cpp (hunk 3 above)
Comment (monkeyman192) [reply scene/gui/tree.cpp:712]: I believe I have fixed the other issues, but I am not sure how I would integrate this into the original method. If it were just called once then it would be easy, but because we need to call it twice in a recursive fashion I am not sure if this is possible? Or maybe I am just blind and/or dumb 😅

File: scene/gui/tree.cpp [hunk 4]
```diff
@@ -709,12 +738,37 @@ TreeItem *TreeItem::get_next_visible(bool p_wrap) {
...
+	return next;
+}
+
 TreeItem *TreeItem::get_child(int p_idx) {
 	_create_children_cache();
 	ERR_FAIL_INDEX_V(p_idx, children_cache.size(), nullptr);
 	return children_cache.get(p_idx);
 }
 
+int TreeItem::get_visible_child_count() {
```
Comment (monkeyman192) [inline scene/gui/tree.cpp:761]: I didn't publicly expose this function, but if it might be useful then I can add it. I can see it having use, but it's also not as efficient for determining whether to conditionally hide something based on whether it has any visible children as simply counting how many children are not hidden as you hide them to avoid doing a second loop.
It still may be useful though in other situations...

File: scene/gui/tree.cpp [hunk 5]
```diff
@@ -682,7 +697,21 @@ TreeItem *TreeItem::get_prev_visible(bool p_wrap) {
 	return current;
 }
 
-TreeItem *TreeItem::get_next_visible(bool p_wrap) {
+TreeItem *TreeItem::get_prev_visible(bool p_wrap) {
+	TreeItem *loop = this;
+	TreeItem *prev = this->_get_prev_visible(p_wrap);
```
Comment (KoBeWi) [inline scene/gui/tree.cpp:702]: It's weird that `_get_prev_visible()` can return a TreeItem that isn't visible. We should rename this method, but it can be done in later PR.

File: scene/gui/tree.cpp [hunk 6]
```diff
@@ -4150,7 +4220,7 @@ int Tree::get_column_minimum_width(int p_column) const {
 					depth += 1;
 				} else {
 					TreeItem *common_parent = item->get_parent();
-					while (common_parent != next->get_parent()) {
+					while (common_parent != next->get_parent() && common_parent) {
```
Comment (KoBeWi) [inline scene/gui/tree.cpp:4223]: Why this change? Doesn't seem related.

File: scene/gui/tree.h [hunk 7]
```diff
@@ -124,6 +124,7 @@ class TreeItem : public Object {
 	Vector<Cell> cells;
 
 	bool collapsed = false; // won't show children
+	bool visible = true; // needs to be drawn
```
Comment (KoBeWi) [inline scene/gui/tree.h:127]: This comment isn't really needed.

File: scene/gui/tree.cpp (hunk 1 above)
Comment (KoBeWi) [reply scene/gui/tree.cpp]: Now I noticed you just copied code from `set_collapsed()`. Seems like that method is wrong too.

File: scene/gui/tree.cpp [hunk 8]
```diff
@@ -544,6 +544,21 @@ bool TreeItem::is_collapsed() {
 	return collapsed;
 }
 
+void TreeItem::set_visible(bool p_visible) {
+	if (visible == p_visible) {
+		return;
+	}
+	visible = p_visible;
+	if (tree) {
+		tree->update();
```
Comment (KoBeWi) [inline scene/gui/tree.cpp:553]: This will update the tree even if a parent item is invisible (so needlessly). I wonder if it's worth to optimize it 🤔

File: doc/classes/TreeItem.xml [hunk 9]
```diff
@@ -687,6 +687,10 @@
 		<member name="disable_folding" type="bool" setter="set_disable_folding" getter="is_folding_disabled">
 			If [code]true[/code], folding is disabled for this TreeItem.
 		</member>
+		<member name="visible" type="bool" setter="set_visible" getter="is_visible">
+			If [code]true[/code], the TreeItem is visible (default).
+			Note that if a TreeItem is set to not be visible, none of its children will be visible either.
```
Comment (KoBeWi) [inline doc/classes/TreeItem.xml:692]:
```suggestion
			If [code]true[/code], the [TreeItem] is visible (default).
			Note that if a [TreeItem] is set to not be visible, none of its children will be visible either.
```

**Review by KoBeWi — APPROVED** [review]
Looks good overall. I left some comments, but any bigger changes can be done in a future PR.

File: scene/gui/tree.cpp (hunk 6 above)
Comment (monkeyman192) [reply scene/gui/tree.cpp:4223]: This is to fix the issue that you raised near the top of this review.
Without this change your example code that didn't work will still not work.
I am not 100% sure why it is needed either, but adding this indeed fixes it.

File: scene/gui/tree.cpp (hunk 8 above)
Comment (monkeyman192) [reply scene/gui/tree.cpp:553]: I am not sure if there is a way to do this without checking parents upwards until one is found to be invisible.
I am guessing that this would still be more efficient than updating the entire tree each time.
If you think it's worth including this in this PR I can do that otherwise we can leave for a future one?

File: scene/gui/tree.cpp (hunk 5 above)
Comment (monkeyman192) [reply scene/gui/tree.cpp:702]: Yeah, I agree. I really couldn't think of a better name 😅

File: scene/gui/tree.cpp (hunk 8 above)
Comment (monkeyman192) [reply scene/gui/tree.cpp:553]: I had some further thought on this and realised I don't think that the `set_collapsed` method has an optimisation like you described above either (so if you collapse or un-collapse a node which is underneath an already collapsed node you would toggle an update of the whole tree.
Maybe it might be worth doing an investigation on some optimisations in a separate ticket/PR.

**Comment by monkeyman192** [comment]
> Looks good overall. I left some comments, but any bigger changes can be done in a future PR.

Thanks very much for the review. I have made the suggested small changes, and have left the others with comments here.
If you are happy with the state of this PR, I am happy for it to be merged and have no other changes to make (unless extra suggestions are made).
Also, let me know if a back-port is required for the 3.x branch.

File: scene/gui/tree.cpp (hunk 6 above)
Comment (KoBeWi) [reply scene/gui/tree.cpp:4220]: Ah, I see why is this needed now. I mean, I know when it could crash without this check, just not sure when this happens. Probably not important 🤔

You could swap the order though, so `common_parent` check is first.
